src/Params/WaveTable.h

```diff
@@ -0,0 +1,659 @@
+
+using Shape1 = Shape<1>;
+using Shape2 = Shape<2>;
+using Shape3 = Shape<3>;
+
+//! Tensor3 that keeps track of its subtensors, using a Tensor of
+//! AbstractRingbuffer
+class Tensor3ForWaveTable : public Tensor<3, wavetable_types::float32>
+{
+    using base_type = Tensor<3, wavetable_types::float32>;
+
+public:
+    Tensor3ForWaveTable(const Shape<3>& shape)
+        : base_type(shape), ringbuffers(shape.dim[1])
+    {
+        for(AbstractRingbuffer& rb : ringbuffers)
+            rb.resize(shape.dim[0]);
+    }
+
+    void swapWith(Tensor3ForWaveTable& other)
+    {
+        base_type::swapWith(other);
+        ringbuffers.swapWith(other.ringbuffers);
+    }
+
+    Tensor<1, AbstractRingbuffer> ringbuffers;
+};
+
+/**
+    All pre-computed data that ADnote needs for generating a note
+    @note This class will only ever reside in the RT thread. Non-RT threads
+          may generate the required tensors, but only send them via bToU.
+          At no time, non-RT threads access content directly.
+*/
+class WaveTable
+{
+public:
+    using float32 = wavetable_types::float32;
+    using IntOrFloat = wavetable_types::IntOrFloat;
+    using WtMode = wavetable_types::WtMode;
+
+    // pure guesses for what sounds good:
+    constexpr const static tensor_size_t num_freqs = 10;
+    constexpr const static tensor_size_t num_semantics = 32; // (in non-WT mode)
+    // number of steps on basefunc knob (the steps are integral)
+    constexpr const static float wt_127f = 127.f;
+    // in WT mod, we have steps of 0.25. This variable adds 3 unused waves padding.
+    constexpr const static tensor_size_t max_semantics_ever = 128 * 4;
+    constexpr const static tensor_size_t num_semantics_wtmod = max_semantics_ever;
+
+    static_assert (num_semantics <= max_semantics_ever, "max_semantics_ever too low");
+    static_assert (num_semantics_wtmod <= max_semantics_ever, "max_semantics_ever too low");
+
+private:
+    //! Semantics for which the waves are built.
+    //! E.g. oscil params or random seed (e.g. 0...127).
+    Tensor1<IntOrFloat> semantics;
+    //! Frequencies for which the waves are built.
+    Tensor1<float32> freqs;
+    //! The waves.
+    //! 1st dim: semantics, 2nd: freqs, 3rd: time.
+    Tensor3ForWaveTable data;
+
+    // Helpers to get the addresses
+    const Tensor1<IntOrFloat>* const semantics_addr = &semantics;
+    const Tensor1<float32>* const freqs_addr = &freqs;
+
+    WtMode m_mode;
+
+    // the following things are just timestamps/markers
+    // - might be merged into less or just one variable
+
+    // Each wavetable request gets a timestamp.
+    // When the wavetable is finished it is saved with that timestamp.
+    // This is useful to mark a wavetable as "outdated", so that we do not
+    // request data for this wavetable, or so that we can ignore outdated
+    // WTs sent.
+    int timestamp_requested = 0, timestamp_current = 0;
```

Comment:
If these timestamps represent time and not just an increment for each change I suspect we want these to be something like int64_t or uint64_t. At least I think I recall the other system timestamps in zyn using those types.

Reply:
Actually we use no clock time anywhere in the whole algorithm, at least from all I remember. Here especially, we start at 0 and keep increasing by 1 (The source is MiddleWare.cpp, `WaveTableRequestHandler::info_t::param_change_time`). So for 32 bit ints, this allows 4 billion parameter changes - which should be enough.

Reply:
hm, then this might just be a naming thing. Instead of it being 'time', perhaps 'step', 'sequence', 'reevaluations', or likely a better word to imply that it's a count. It's a comparatively low priority issue now that it's been clarified.